components/iam/pkg/oidc/oauth2.go

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package oidc
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"golang.org/x/oauth2"
+)
+
+type OAuth2Result struct {
+	OAuth2Token *oauth2.Token
+	Redirect    string
+}
+
+type keyOAuth2Result struct{}
+
+func OAuth2Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		log.Trace("at oauth2 middleware")
+		ctx := r.Context()
+		config, ok := ctx.Value(keyOIDCClientConfig{}).(OIDCClientConfig)
+		if !ok {
+			http.Error(rw, "config not found", http.StatusInternalServerError)
+			return
+		}
+
+		// http-only cookie written during flow start request
+		stateCookie, err := r.Cookie(stateCookieName)
+		if err != nil {
+			http.Error(rw, "state cookie not found", http.StatusBadRequest)
+			return
+		}
+		// the starte param passed back from IdP
+		stateParam := r.URL.Query().Get("state")
+		if stateParam == "" {
+			http.Error(rw, "state param not found", http.StatusBadRequest)
+			return
+		}
+		// on mismatch, obviously there is a client side error
+		if stateParam != stateCookie.Value {
+			http.Error(rw, "state did not match", http.StatusBadRequest)
+			return
+		}
+
+		code := r.URL.Query().Get("code")
+		if code == "" {
+			http.Error(rw, "code param not found", http.StatusBadRequest)
+			return
+		}
+
+		oauth2Token, err := config.OAuth2Config.Exchange(ctx, code)
+		if err != nil {
+			http.Error(rw, "failed to exchange token: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		ctx = context.WithValue(ctx, keyOAuth2Result{}, OAuth2Result{
+			OAuth2Token: oauth2Token,
+		})
+		next.ServeHTTP(rw, r.WithContext(ctx))
+	})
+}
```

Comment:
I'd strongly recommend not to over-use middlewares which extract concrete data. Here's why:

* Golang's context is not the same as the `express.Request` context. The context in go is used for downstream calls irrespective of whether it's an http handler or not. Using the context to pass request specific arguments results in carry extra baggage.
* The middleware forces you write/and parse the data out of the context. Because context is not type safe, you incur extra overhead in extracting the data, but also risk cases where the context does not parse correctly.
* It forces you to handle the existence of the data twice, once on the middleware here, and another when you pull the value from the context
* Middlewares tend to make sense, once you have general purpose logic you want to re-use across handlers. Here, we only have a single handler. This causes a couple of problems:
	* Testing the handler becomes harder, you always have to test it together with the middleware
	* Readability becomes harder, the reader now has to step a layer above and figure out what's populating the context 
* It makes handling errors from context values much harder. The context of why the value is not present is missing. Yes, the middleware should return an error but we need to be defensive and also guard this in the handler.

Reply:
That said, checking for a solid connection should be a must here, thus it's worth revisiting tests. 👍🏻